src/compiler/checker.ts

```diff
@@ -16549,6 +16561,40 @@ namespace ts {
             return unknownType;
         }
 
+        // Gets the type of object literal or array literal of destructuring assignment.
+        // { a } from 
+        //     for ( { a } of elems) {
+        //     }
+        // [ a ] from 
+        //     [a] = [ some array ...]
+        function getTypeOfArrayLiteralOrObjectLiteralDestructuringAssignment(expr: Expression): Type {

```

Comment:
we should then change the entry point to return a targetSymbol for a bindingPattern instead of a type.